sycamore/writers/file_writer.py

```diff
@@ -0,0 +1,155 @@
+        prefer_text: bool = True,
+        **ray_remote_args,
+    ):
+        """Initializes a FileWriter instance.
+
+        Args:
+            plan: A Sycamore plan representing the DocSet to write out.
+            path: The path prefix to write to. Should include the scheme.
+            filesystem: The pyarrow.fs FileSystem to use.
+            filename_fn: A function for generating a file name. Takes a Document
+                and returns a unique name that will be appended to path.
+            prefer_text: If true will use text_representation if both text and binary
+                representations are set.
+            ray_remote_args: Arguments to pass to the underlying execution environment.
+        """
+
+        super().__init__(plan, **ray_remote_args)
+        self.path = path
+        self.filesystem = filesystem
+        self.filename_fn = filename_fn
+        self.prefer_text = prefer_text
+        self.ray_remote_args = ray_remote_args
+
+    def execute(self) -> Dataset:
+        dataset = self.child().execute()
+
+        dataset.write_datasource(
+            _WritableFilePerRowDataSource(),
+            path=self.path,
+            dataset_uuid=uuid.uuid4(),
+            filesystem=self.filesystem,
+            filename_fn=self.filename_fn,
+            prefer_text=self.prefer_text,
+            ray_remote_args=self.ray_remote_args,
+        )
+
+        return dataset
+
+
+# Some of this code is taken from Ray's FileBasedDatasource. We should switch to using that
+# Once it supports using a custom filename.
+class _WritableFilePerRowDataSource(FileBasedDatasource):
+    _WRITE_FILE_PER_ROW = True
+
+    def write(
+        self,
+        blocks: Iterable[Block],
+        ctx: TaskContext,
+        path: str,
+        filesystem: Optional[FileSystem] = None,
+        filename_fn: Callable[[Document], str] = default_filename,
+        prefer_text: bool = True,
+        **write_args,
+    ) -> WriteResult:
+        from ray.data.datasource.file_based_datasource import (
+            _open_file_with_retry,
+            _resolve_paths_and_filesystem,
+            _unwrap_s3_serialization_workaround,
+        )
+
+        path, filesystem = _resolve_paths_and_filesystem(path, filesystem)
+        path = path[0]
+
+        for block in blocks:
+            block = BlockAccessor.for_block(block)  # .to_arrow().to_pylist()
+
+            fs = _unwrap_s3_serialization_workaround(filesystem)
+
+            for row in block.iter_rows(public_row_format=True):
+                doc = Document(row)
+                filename = filename_fn(doc)
+
+                write_path = os.path.join(path, filename)
+                logger.debug(f"Writing file at {write_path}")
+
+                # with fs.open_output_stream(write_path) as outfile:
+                with _open_file_with_retry(write_path, lambda: fs.open_output_stream(write_path)) as outfile:
+                    if prefer_text and doc.text_representation is not None:
```

Comment:
is there case, we want write both text and binary into same file, e.g an image? or is there a consideration for writing based on document type instead of a flag?

Reply:
Good question. This also completely ignores properties. Maybe calling this `files` is too expansive, since it's really just writing the main content. I think we would need a different writer for something like JSON, even if it still goes to a file. 

Agreed on document type. Maybe we have some kind of is_text/is_binary flag that we can use to distinguish? Not sure if we wan to add that in this PR or wait until we have a more holistic approach.